### PR TITLE
4. As a user, I want to set up a new shopping list so I can start tracking purchased items.

### LINKED ISSUE #4: 4. As a user, I want to set up a new shopping list so I can start tracking purchased items.
## Summary 

A shopping list is a set of items associated with the owner's unique id and the name of the shopping list. Users need to be able to create new shopping lists. To do this, the `Home` view should present them with a form that allows them to enter the name of their list. Additionally, we need to do a few things to create that list:

1. Save the token to `localStorage`
2. Save the list to our database using the `createList` function in `firebase.js` 
3. Show the user the list view
## Acceptance criteria
UI-related tasks:
- [x] The `Home` view displays a form that allows users to enter the name of a shopping list and then create a list with that name.
- [x] The input that accepts the name of the item has a semantic `label` element associated with it
- [x] The user can submit this form with both the mouse and the `Enter` key  
- [x]  When the user submits the form, they see a message indicating that the list either was or was not created and saved to the database.

Data-related tasks:
- [x] Clicking the button uses the `createList` function in `src/api/firebase.js` to create a new shopping list for the user.
- [x] The shopping list path is stored in local storage using the `setListPath` function.
- [x] Once the list has been created and saved, the user is redirected to the `List` view.

## Notes
- The `createList` function takes **three** arguments: the current user's id, the current user's email, and the name of the shopping list.
- The `setListPath` function takes **one** argument: the new list token.
- If you get stuck trying to redirect the user, check out the [useNavigate](https://reactrouter.com/en/6.14.2/hooks/use-navigate) hook from `react-rotuer`.

### PR DESCRIPTION
**Pull request by 3campos**
## Description

This code creates an accessible form that allows users to create grocery lists which are added to the Firestore database. When the user creates a list, a message appears confirming whether the creation was successful or not. It also adds the path to local storage. After creating a list, the app navigates users to the list page.

## Related Issue

closes #4 

## Acceptance Criteria

UI-related tasks:

[x] The Home view displays a form that allows users to enter the name of a shopping list and then create a list with that name.
[x] The input that accepts the name of the item has a semantic label element associated with it
[x] The user can submit this form with both the mouse and the Enter key
[x] When the user submits the form, they see a message indicating that the list either was or was not created and saved to the database.

Data-related tasks:

[x] Clicking the button uses the createList function in src/api/firebase.js to create a new shopping list for the user.
[x] The shopping list path is stored in local storage using the setListPath function.
[x] Once the list has been created and saved, the user is redirected to the List view.

### After

https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/105468537/5b91898a-b820-4262-b445-95b5507620bb

## Testing Steps / QA Criteria

1. From the code, type "npm start" in your terminal
2. Click the Sign In button and sign in using the user you have connected to firebase
3. Type a grocery list name
4. Click "Create List" or press "Enter" on your keyboard
5. Open your DevTools Window 
6. Navigate to "Application" tab and open the drop down menu for "Local Storage"
7. Check that the path to the list is added into local storage for your localhost page

### CODE REVIEW

**Comment by StefieCaff** [comment]
@3campos @eonflower @hun-ah 🤷🏾‍♀️🤷🏻Thinking about if the listName state needs to be cleared anywhere, or if that is already done somewhere in the depth of Firebase.

**Comment by eonflower** [comment]
A couple notes I have:
- When I create a list, it adds that list name to the list, but I have gotten "Your list (list name) was not created, please try again" regardless of the actual outcome
- I am not being redirected to the list page when I create a list
- In the CreateList function in our firebase file, the parameters are userId, userEmail, and listName, but in the home view you have user.displayName. I think this could create problems down the line if their are users with the same name.
- For conditionally rendering the message, I have read that using && can run into some issues, so we might want to use a ternary operator instead??? (article with more info: https://dev.to/maafaishal/avoid-operator-for-conditional-rendering-in-react-2de)

I saw that y'alls redirects you to the correct page, so I am curious why mine is not 🤔 let me know if you need any clarification on my notes!

**Review by hun-ah — COMMENTED** [review]
Nice work! The code is working well for me, the only thing I am wondering about is after I successfully create a list, I am redirected to the list page but I'm not seeing the message indicating I have successfully created the list. I assume this is because it is instantly navigating to the list page. You may want to use a setTimeout function to display the message first and then navigate to the list page.

**Comment by eonflower** [comment]
Here is what I see

https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/98231995/084815a1-12cf-42d7-8531-4b882340ee75

**Comment by 3campos** [comment]
> * In the CreateList function in our firebase file, the parameters are userId, userEmail, and listName, but in the home view you have user.displayName. I think this could create problems down the line if their are users with the same name.

@eonflower, thanks for the feedback! Would you git pull our branch again? We initially used the "user.displayName" code but removed it long before our most recent commit and push today (which was done before our pull request). I suspect that the other problems that you encountered may be a result of the older commit being pulled onto your machine.  Please let @StefieCaff and me know if this resolves these issues or if you encounter the same problems. Thanks!

**Comment by eonflower** [comment]
@3campos Okay yes! That worked. I did a git pull when I first opened your branch, but maybe it didn't work??? Though, I also don't know how it would have gotten your old commits because I was not on your branch at all until now 😂 

All this is to say, it works now lol.

**Comment by 3campos** [comment]
> You may want to use a setTimeout function to display the message first and then navigate to the list page.

@hun-ah Thanks for catching this! When I started reading through your feedback, I thought "setTimeout will probably solve this" 🙌 . I implemented this method to delay the automatic navigation to the link page by 2 seconds. It works like a charm now! 

@StefieCaff, this resolves Hannah's feedback. 😄

**Review by StefieCaff — APPROVED** [review]

**Comment by StefieCaff** [comment]
@3campos I think that we need to update the conditional rendering that @eonflower mentioned... also @hun-ah I think we can all take a look possible solutions at the bottom-- there is one that adds an esLint rule that we could consider.

I was also considering making a new component like ya'll did for the addItem.jsx. If you all agree, I would like to refactor our form into it's own component  addList.jsx -- I wasn't sure because the app seemed uncomplicated and small enough-- but I like the idea of keeping Home.jsx cleaner as well. What do you think?

**Comment by hun-ah** [comment]
@3campos looks awesome!

@eonflower good catch on the conditional rendering, could be worth making the switch to a ternary operator to be safe.

@StefieCaff I'm okay with you creating the `addList` component and moving the form logic into it, I agree it will help keep `Home` clean.

**Comment by StefieCaff** [comment]
created new component and updated message rendering
[3773149](https://github.com/the-collab-lab/tcl-66-smart-shopping-list/pull/22/commits/377314933da04d1cb088edd0c6204bca69e76a81)

[f8536eb](https://github.com/the-collab-lab/tcl-66-smart-shopping-list/pull/22/commits/f8536eb4c98a5749dac2c3a035395416dce5f0f1)

**Review by eonflower — APPROVED** [review]
Looks great! I like that you made the AddList into a component. Keeps the home view nice and clean :)

**Review by hun-ah — APPROVED** [review]
Looks good!

File: src/components/AddList.jsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+export default function AddList({ setListPath }) {
+	const [listName, setListName] = useState('');
+	const [message, setMessage] = useState('');
+	const navigate = useNavigate();
+	const { user } = useAuth();
+
+	const handleSubmit = (event) => {
+		event.preventDefault();
+		try {
+			const newList = createList(user.uid, user.email, listName);
```
Comment (deeheber) [inline src/components/AddList.jsx:15]: Not blocking: you could check if a user is present (aka if they're logged in or not) and throw up an alert prompting the user to login to create a new list if they aren't logged in in this handleSubmit function.

**Review by deeheber — APPROVED** [review]
This looks good to merge to me.

Had the one non-blocking suggestion since it wasn't part of the requirements, it's totally up to you if you want to add that in or not before merging. 🚀

File: src/components/AddList.jsx (hunk 1 above)
Comment (StefieCaff) [reply src/components/AddList.jsx:15]: Good idea :) I have a question though, what does it mean not blocking?

Also, we had casually discussed adding functionality later--- like private vs public routes or validation of some kind. If we do that, would adding this to the handleSubmit function be duplicitous?

Comment (deeheber) [reply src/components/AddList.jsx:15]: Update: we chatted about this via DM. Feel free to ignore that comment for now and merge when ready, I missed the convo about adding this later.

**Pull request merged**